File: neetcode150/linked_lists/leetcode138.py
from typing import Optional

from data_structures.linked_list import RandomNode as Node


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        copies = {None: None}

        curr = head

        while curr:
            copies[curr] = Node(curr.val)
            curr = curr.next

        curr = head

        while curr:
            copies[curr].next = copies[curr.next]
            copies[curr].random = copies[curr.random]

            curr = curr.next

        return copies[head]

File: js_algorithms/neetcode150/linked_lists/leetcode138.js
function Node(val, next, random) {
  this.val = val;
  this.next = next;
  this.random = random;
}

/**
 * @param {Node} head
 * @return {Node}
 */
const copyRandomList = function (head) {
  const mp = new Map();
  mp.set(null, null);
  let curr = head;

  while (curr) {
    // keep track of node copies
    mp.set(curr, new Node(curr.val));
    curr = curr.next;
  }

  curr = head;

  while (curr) {
    const node = mp.get(curr);
    if (node) {
      node.next = mp.get(curr?.next);
      node.random = mp.get(curr?.random);
    }

    curr = curr.next;
  }

  return mp.get(head);
};

module.exports = { copyRandomList };
